src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
```

Comment:
WithKey and WithName look nice, but we've no means of enforcing that they actually conform to the previously added item.  The below will compile:

services.AddTransient<Zero>().WithKey<int, IValue<string>, That>("that");

Which doesn't really make sense.  Do we want to be more explicit about what we're doing here, at the cost of readability?  Something like:

services.AddTransient<Zero>();
services.AddKeyForService<IValue<string>, Zero, int>(0);